zallet/src/components/json_rpc/methods/get_address_for_account.rs

```diff
@@ -0,0 +1,208 @@
+                            "Error: P2SH addresses can not be created using this RPC method.",
+                        ))
+                    }
+                    "p2pkh" => p2pkh = ReceiverRequirement::Require,
+                    _ => invalid_receivers.push(receiver_type),
+                }
+            }
+
+            if invalid_receivers.is_empty() {
+                UnifiedAddressRequest::custom(orchard, sapling, p2pkh)
+                    .map(|request| (receiver_types, request))
+                    .map_err(|_| {
+                        LegacyCode::InvalidParameter.with_static(
+                            "Error: cannot generate an address containing no shielded receivers.",
+                        )
+                    })
+            } else {
+                Err(LegacyCode::InvalidParameter.with_message(format!(
+                    "{:?} {}. Arguments must be “p2pkh”, “sapling”, or “orchard”",
+                    // TODO: Format nicely (matching zcashd).
+                    invalid_receivers,
+                    if invalid_receivers.len() == 1 {
+                        "is an invalid receiver type"
+                    } else {
+                        "are invalid receiver types"
+                    }
+                )))
+            }
+        }
+        _ => {
+            // zcashd default is the best and second-best shielded receiver types, and the
+            // transparent (P2PKH) receiver type. That currently corresponds to all possible
+            // receiver types.
+            let request = UnifiedAddressRequest::unsafe_custom(
+                ReceiverRequirement::Require,
+                // TODO: Determine whether `zcashd` used `Require` or `Allow` semantics
+                // when specifying a diversifier index.
+                ReceiverRequirement::Require,
+                ReceiverRequirement::Require,
+            );
+            Ok((
+                vec!["orchard".into(), "sapling".into(), "p2pkh".into()],
+                request,
+            ))
+        }
+    }?;
+
+    let diversifier_index = diversifier_index.map(parse_diversifier_index).transpose()?;
+
+    let (address, diversifier_index) = if let Some(diversifier_index) = diversifier_index {
+        match wallet
+            .get_address_for_index(account_id, diversifier_index, request)
+            .map_err(|e| map_sqlite_error(e, &account))?
+        {
+            Some(address) => Ok((address, diversifier_index)),
+            // zcash_client_sqlite only returns `Ok(None)` if the diversifier index is
+            // invalid for one of the requested receiver types.
+            None => Err(LegacyCode::Wallet.with_message(format!(
+                "Error: diversifier index {} cannot generate an address with the requested receivers.",
+                u128::from(diversifier_index),
+            ))),
+        }
+    } else {
+        wallet
+            .get_next_available_address(account_id, request)
```

Comment:
After the changes in https://github.com/zcash/librustzcash/pull/1673, this will not behave as described in the documentation (it will not necessarily produce the next diversifier index in sequence unless the request specifies a transparent receiver.). If the sequentiality of shielded-only diversifier indices is considered essential, then we'll have to add a `force_sequential_diversifier_index_allocation` flag or something of the sort to this API.

Reply:
One could argue that "valid for the list of receiver types" is sufficient flexibility to let us switch to time-based for shielded-only. But we probably should avoid changing these particular semantics (unless we want to do outreach to the ecosystem participants who have said they are using this API, to determing what aspect of this guarantee they depend on).

Maybe the way to handle this is at time of constructing the `WalletDb`? If we don't have a clock, we can't do time-based address derivation. Then it could be a Zallet config option for whether we allow time-based derivation or not. (Is there anything else we might want the clock for in future, where we couldn't just treat clock presence as equivalent to time-based derivation?)

Reply:
Any form of decorrelation to reduce timing metadata leakage might need a clock. I'd prefer making that an explicit flag (or however we handle other config options).

Reply:
Part of the issue I had hoped to avoid with the time-based address generation is the question of what to do if a caller requests > GAP_LIMIT shielded-only addresses in a row. If we "use up" transparent child indices with shielded address derivation, then we'll end up in a situation where someone's desire to generate a new shielded-only address conflicts with the transparent gap limit. 

Suffice it to say that I think we're better off with the new `librustzcash` behavior, and should just document it.